doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1257,409 +1261,344 @@ Managing the local context
-   :name: enough
+   Implements the non-dependent case of the :ref:`App <app_rule>` typing rule,
+   the Modus Ponens inference rule.  It is equivalent to
+   :n:`enough (@ident: @one_term). revert @ident.`
+   This tactic is generally considered obsolete but it is still widely
+   used in old scripts.
 
-   This adds a new hypothesis of name :token:`ident` asserting :token:`type` to the
-   goal the tactic :tacn:`enough` is applied to. A new subgoal stating :token:`type` is
-   inserted after the initial goal rather than before it as :tacn:`assert` would do.
+.. tacn:: pose proof @term {? @as_ipat }
+          pose proof ( @ident := @term )
+   :name: pose proof; _
 
-.. tacv:: enough @type
+   The first form behaves like :n:`assert @type {? @as_ipat } by exact @term`
+   where :token:`type` is the type of :token:`term`.
 
-   This behaves like :n:`enough (@ident : @type)` with the name :token:`ident` of
-   the hypothesis generated by Coq.
+   .. Théo notes it's odd that the first form uses @term instead of @one_term
 
-.. tacv:: enough @type as @simple_intropattern
+   The second form is equivalent to :n:`assert (@ident := @term)`.
 
-   This behaves like :n:`enough @type` using :token:`simple_intropattern` to name or
-   destruct the new hypothesis.
+   .. tacn:: epose proof @term {? @as_ipat }
+             epose proof ( @ident := @term )
+      :name: epose proof; _
 
-.. tacv:: enough (@ident : @type) by @tactic
-          enough @type {? as @simple_intropattern } by @tactic
+      While :tacn:`pose proof` expects that no existential variables are generated by
+      the tactic, :tacn:`epose proof` removes this constraint.
 
-   This behaves as above but with :token:`tactic` expected to solve the initial goal
-   after the extra assumption :token:`type` is added and possibly destructed. If the
-   :n:`as @simple_intropattern` clause generates more than one subgoal, :token:`tactic` is
-   applied to all of them.
+.. tacn:: specialize @one_term_with_bindings {? as @simple_intropattern }
+   :name: specialize
 
-.. tacv:: eenough @type {? as @simple_intropattern } {? by @tactic }
-          eenough (@ident : @type) {? by @tactic }
-   :name: eenough; _
-
-   While the different variants of :tacn:`enough` expect that no existential
-   variables are generated by the tactic, :tacn:`eenough` removes this constraint.
-
-.. tacv:: cut @type
-   :name: cut
-
-   This tactic applies to any goal. It implements the non-dependent case of
-   the “App” rule given in :ref:`typing-rules`. (This is Modus Ponens inference
-   rule.) :n:`cut U` transforms the current goal :g:`T` into the two following
-   subgoals: :g:`U -> T` and :g:`U`. The subgoal :g:`U -> T` comes first in the
-   list of remaining subgoal to prove.
-
-.. tacv:: specialize (@ident {* @term}) {? as @simple_intropattern}
-          specialize @ident with @bindings {? as @simple_intropattern}
-   :name: specialize; _
+   .. todo PR: Where is @ident now?  Need to adapt
```

Comment:
What is referred to by `ident` here is the head constant of `@one_term_with_bindings`. Technically, it is not limited to `ident`, a `qualid` is supported here (and even forms where there is no head constant, but I'm not exactly sure how it works in this case).

Reply:
When the head term is a lambda the binders are generalized over in the new hypothesis.  So `specialize (fun x y => H y x)` will swap the arguments of `H`.  I believe in other cases (headed by global constant, headed by `match`, etc) `specialize` will either error, or else act as either `pose proof` or `generalize`, though I don't recall which.  Can someone check?

Reply:
No, if the head is a global constant, it behaves as documented, i.e., close to `pose proof` except that it adds a premise to the goal if no `as` clause is provided.

Reply:
Ah, I guess that's why I couldn't recall if it was like pose proof or generalize; it's in between.  How confident are we that this behavior is intended rather than a bug in the code?  (Also, I expect global constants are handled the same way as things that are neither constants nor lambdas.  Though I guess I don't know whether or not let ins are reduced.)

Reply:
> How confident are we that this behavior is intended rather than a bug in the code?

A long time ago, it was quite common to add a premise to the goal (see elim and co.), so I would be surprised that this was not intentional.